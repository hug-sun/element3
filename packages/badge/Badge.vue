<template>
  <div class="el-badge">
    <slot></slot>
    <transition name="el-zoom-in-center">
      <sup
        v-show="!hidden && (content || content === 0 || isDot)"
        v-text="content"
        class="el-badge__content"
        :class="[
          'el-badge__content--' + type,
          {
            'is-fixed': $slots.default,
            'is-dot': isDot
          }
        ]"
      >
      </sup>
    </transition>
  </div>
</template>

<script>
import { computed, toRefs } from 'vue'

export default {
  name: 'ElBadge',

  props: {
    value: [String, Number],
    max: Number,
    isDot: Boolean,
    hidden: Boolean,
    type: {
      type: String,
      validator(val) {
        return (
          ['primary', 'success', 'warning', 'info', 'danger'].indexOf(val) > -1
        )
      }
    }
  },

  setup(props) {
    const { isDot, max, value } = toRefs(props)
    const content = useContent(isDot, max, value)

    return {
      content
    }
  }
}

const useContent = (isDot, max, value) => {
  const content = computed(() => {
    if (isDot.value) return

    if (
      max &&
      typeof value.value === 'number' &&
      typeof max.value === 'number'
    ) {
      return max.value < value.value ? `${max.value}+` : value.value
    }

    return value.value
  })

  return content
}
</script>
