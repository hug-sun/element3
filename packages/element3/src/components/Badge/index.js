import ElBadge from './src/Badge.vue'

/* istanbul ignore next */
ElBadge.install = function (app) {
  app.component(ElBadge.name, ElBadge)
}

export { ElBadge }
