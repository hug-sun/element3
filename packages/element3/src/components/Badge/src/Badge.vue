<template>
  <div class="el-badge">
    <slot></slot>
    <transition name="el-zoom-in-center">
      <sup
        v-if="isShow"
        class="el-badge__content"
        v-text="badgeNumber"
        :class="[
          'el-badge__content--' + type,
          {
            'is-fixed': $slots.default,
            'is-dot': isDot
          }
        ]"
      ></sup>
    </transition>
  </div>
</template>

<script>
import { defineComponent, computed } from 'vue'
import { isString } from '../../../utils/types'

export default defineComponent({
  name: 'ElBadge',

  props: {
    value: [String, Number],
    max: Number,
    isDot: Boolean,
    hidden: Boolean,
    type: {
      type: String,
      default: 'danger',
      validator(val) {
        return ['primary', 'success', 'warning', 'info', 'danger'].includes(val)
      }
    }
  },

  setup(props) {
    const badgeNumber = useBadgeNumber(props)
    const isShow = useShow(props)
    return {
      badgeNumber,
      isShow
    }
  }
})

const useBadgeNumber = (props) => {
  return computed(() => {
    if (!props.max || isString(props.value)) {
      return props.value
    }

    if (props.value <= props.max) {
      return props.value
    }

    return `${props.max}+`
  })
}

const useShow = (props) => {
  return computed(() => {
    if (props.hidden) {
      return false
    }
    if (props.value == 0) {
      return false
    }
    return true
  })
}
</script>
