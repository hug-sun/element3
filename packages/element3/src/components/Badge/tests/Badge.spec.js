import Badge from '../src/Badge.vue'
import { mount } from '@vue/test-utils'

describe('Badge', () => {
  describe('props', () => {
    it('should render props.value', () => {
      const wrapper = mount(Badge, {
        props: {
          value: 13
        }
      })

      expect(wrapper.find('.el-badge__content').text()).toContain('13')
    })

    it('should render props.value when value lt max', () => {
      const wrapper = mount(Badge, {
        props: {
          value: 13,
          max: 20
        }
      })

      expect(wrapper.find('.el-badge__content').text()).toContain('13')
    })

    it('should render string props.value', () => {
      const wrapper = mount(Badge, {
        props: {
          value: 'new~'
        }
      })

      expect(wrapper.find('.el-badge__content').text()).toContain('new~')
    })

    it('should not render when props.value eq 0', () => {
      const wrapper = mount(Badge, {
        props: {
          value: 0
        }
      })

      expect(wrapper.find('.el-badge__content').exists()).toBeFalsy()
    })

    it('should  render solt', () => {
      const wrapper = mount(Badge, {
        props: {
          value: 0
        },
        slots: {
          default() {
            return 'default slot'
          }
        }
      })

      expect(wrapper.find('.el-badge').text()).toContain('default slot')
    })

    it('should render `+` over max value', () => {
      const wrapper = mount(Badge, {
        props: {
          value: 13,
          max: 10
        }
      })

      expect(wrapper.find('.el-badge__content').text()).toContain('10+')
    })

    it('should render dot', () => {
      const wrapper = mount(Badge, {
        props: {
          value: 13,
          max: 10,
          isDot: true
        }
      })

      expect(wrapper.find('.is-dot').text()).toBeTruthy()
    })

    it('should not render when hidden', () => {
      const wrapper = mount(Badge, {
        props: {
          value: 13,
          max: 10,
          hidden: true
        }
      })

      expect(wrapper.find('sup').exists()).toBeFalsy()
    })

    it('should  render props.type', () => {
      const wrapper = mount(Badge, {
        props: {
          value: 13,
          max: 10,
          type: 'primary'
        }
      })

      expect(wrapper.find('sup').classes()).toContain(
        'el-badge__content--primary'
      )
    })
  })
})
