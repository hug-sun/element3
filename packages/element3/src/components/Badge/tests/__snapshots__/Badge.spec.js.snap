// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Badge snapshot 1`] = `
<div
  class="el-badge"
>
  
  
  <transition-stub>
    <sup
      class="el-badge__content el-badge__content--danger"
    />
  </transition-stub>
</div>
`;
